src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -0,0 +1,264 @@
+                {
+                    AddToList(elements, W3CLoggingFields.UriStem, request.Path.Value);
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.UriQuery))
+                {
+                    AddToList(elements, W3CLoggingFields.UriQuery, request.QueryString.Value);
+                }
+
+                if ((W3CLoggingFields.RequestHeaders & options.LoggingFields) != W3CLoggingFields.None)
+                {
+                    var headers = request.Headers;
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.Host))
+                    {
+                        if (headers.TryGetValue(HeaderNames.Host, out var host))
+                        {
+                            AddToList(elements, W3CLoggingFields.Host, host.ToString());
+                        }
+                    }
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.Referer))
+                    {
+                        if (headers.TryGetValue(HeaderNames.Referer, out var referer))
+                        {
+                            AddToList(elements, W3CLoggingFields.Referer, referer.ToString());
+                        }
+                    }
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.UserAgent))
+                    {
+                        if (headers.TryGetValue(HeaderNames.UserAgent, out var agent))
+                        {
+                            AddToList(elements, W3CLoggingFields.UserAgent, agent.ToString());
+                        }
+                    }
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.Cookie))
+                    {
+                        if (headers.TryGetValue(HeaderNames.Cookie, out var cookie))
+                        {
+                            AddToList(elements, W3CLoggingFields.Cookie, cookie.ToString());
+                        }
+                    }
+                }
+            }
+
+            var response = context.Response;
+
+            try
+            {
+                await _next(context);
+            }
+            catch
+            {
+                // Write the log
+                if (_hasLogged)
+                {
+                    _w3cLogger.Log(elements);
+                }
+                throw;
+            }
+
+            if (options.LoggingFields.HasFlag(W3CLoggingFields.UserName))
+            {
+                AddToList(elements, W3CLoggingFields.UserName, context?.User?.Identity?.Name ?? "");
+            }
+
+            if (options.LoggingFields.HasFlag(W3CLoggingFields.ProtocolStatus))
+            {
+                AddToList(elements, W3CLoggingFields.ProtocolStatus, response.StatusCode.ToString(CultureInfo.InvariantCulture));
+            }
+
+            if ((W3CLoggingFields.ResponseHeaders & options.LoggingFields) != W3CLoggingFields.None)
```

Comment:
Server isn't a response header, it's machine state, this should be moved before await _next.

ResponseHeaders can be removed.

Reply:
Nice catch, thanks